report_generation/reports/frequency_reports.py

```diff
@@ -1,62 +1,123 @@
         """
+        thin = Side(border_style="thin", color="000000")
+        border = Border(top=thin, left=thin, right=thin, bottom=thin)
+
         color_mapping = {
-            'Loaded for Delivery': '#00aeed',               # Blue
-            'POD Details Captured': '#d0e833',              # Green
-            'POD Image Scanned': '#d0e833',
-            'Attempted delivery': '#f7bc00',                # Orange
-            'Checked in at Origin Depot': '#f7bc00',
-            'Consignment details captured': '#f7bc00',
-            'Event Scan Blocked': '#f7bc00',
-            'Floor check': '#f7bc00',
-            'Inbound Manifest': '#f7bc00',
-            'Manifest Transferred': '#f7bc00',
-            'Mis-routed': '#f7bc00',
-            'Received at origin depot': '#f7bc00',
-            'Remove from manifest/tripsheet': '#f7bc00',
-            'Return to Client': '#f7bc00',
-            'Return to Depot': '#f7bc00',
-            'Reverse logistics floor check': '#f7bc00',
-            'Swadded': '#f7bc00',
-            'Unload manifest/tripsheet': '#f7bc00',
-            'Floor check - Depot collection': '#eac7e6',    # Purple
-            'Chain store floor check': '#fdf900',            # Yellow
-            'Floor check - Booking cargo': '#fdf900',
-            'Outbound Manifest Load': '#fdf900',
-            'Other': 'White'  # Default color for other entries
+            "Attempted delivery": "f7bc00",
+            "Chain store floor check": "fdf900",
+            "Checked in at Origin Depot": "f7bc00",
+            "Consignment details captured": "f7bc00",
+            "Event Scan Blocked": "f7bc00",
+            "Floor check": "f7bc00",
+            "Floor check - Booking cargo": "fdf900",
+            "Floor check - Depot collection": "eac7e6",
+            "Inbound Manifest": "f7bc00",
+            "Loaded for Delivery": "00aeed",
+            "Manifest Transferred": "f7bc00",
+            "Mis-routed": "f7bc00",
+            "Outbound Manifest Load": "fdf900",
+            "POD Details Captured": "d0e833",
+            "POD Image Scanned": "d0e833",
+            "Received at origin depot": "f7bc00",
+            "Remove from manifest/tripsheet": "f7bc00",
+            "Return to Client": "f7bc00",
+            "Return to Depot": "f7bc00",
+            "Reverse logistics floor check": "f7bc00",
+            "Swadded": "f7bc00",
+            "Unload manifest/tripsheet": "f7bc00",
+            "Other": "FFFFFF",  # Assuming white for "Other"
         }
 
-        return df.style.apply(
-                lambda row: [f'background-color: {color_mapping.get(row["Last Event"], "white")}'] * len(row),
-                axis=1
-            )
+        # Get the index of the 'Last Event' column dynamically
+        last_event_col_index = df.columns.get_loc("Last Event")
+
+        for r_idx, row in enumerate(dataframe_to_rows(df, index=False, header=True), start=start_row):
```

Comment:
Tash, is this for the styling? Weren't we using an apply? I want us to avoid looping through all the rows of all the files please

Reply:
The apply function became redundant, we can call and chat about it in person but the colour mapping now takes place in the _append_df_to_sheet function (because coloring/ styling has to be the last action applied to a df). I can look into removing the loop but it doesn't actually add any overhead processing time that I could see